### PR TITLE
Expose instance uploader activity to external apps

### LINKED ISSUE #2454: Request ODK Collect to submit specific form instances using custom credentials
#### Software and hardware versions 
Collect v1.16.1, Android v6.0.1, Samsung Galaxy J5

#### Problem description

Currently there is no way to request/command ODK Collect to upload specific form instances using either default settings OR custom url/username/password.

#### Expected behavior
An external app should be able to request ODK Collect to submit form instances through an intent. 
This should be done in the foreground using the current activities available

Priority of submission_url to be used -> Least priority is settings, then form, then intent

### PR DESCRIPTION
**Pull request by ekigamba**
Closes #2454 

#### What has been done to verify that this works as intended?

This has been tested with ODK Collect Intents tester [here](https://github.com/grzesiek2010/collectTester/pull/22) on: 
1. Samsung Galaxy Tab S2 Android v7.0
2. Infinix Zero 3 Android v6.0

#### Why is this the best possible solution? Were any other approaches considered?
It is takes the least effort & its the only way to do this in the foreground while reusing the current codebase.

#### Are there any risks to merging this code? If so, what are they?
No

#### Do we need any specific form for testing your changes? If so, please attach one.
No

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew pmd checkstyle lintDebug spotbugsDebug` and confirmed all checks still pass.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

## Implementation:

An external app has to call the `InstanceUploaderActivity` and pass the instance ids of the instances they want uploaded. This starts the activity.

Other optional extras that can be passed is the *submission url*, *username* and *password*.

The activity has no result since the external app can access the Content Provider to get the new states of the instance. Any error message will be displayed by the InstanceUploaderActivity and are already present.

The url, username & password overrides all previous settings & submission url in the form.

The following is the sample code for making such an intent call:

```
Intent intent = new Intent();
intent.setClassName("org.odk.collect.android", "org.odk.collect.android.activities.InstanceUploaderActivity");
intent.putExtra("instances", new long[]{5, 9, 12});
intent.putExtra("URL", "https://opendatakit.appspot.com");
intent.putExtra("USERNAME", "user");
intent.putExtra("PASSWORD", "mypassword");
startActivity(intent);
```

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java [hunk 1]
```diff
@@ -93,14 +96,42 @@ protected void onCreate(Bundle savedInstanceState) {
         long[] selectedInstanceIDs = null;
         if (savedInstanceState != null && savedInstanceState.containsKey(TO_SEND)) {
             selectedInstanceIDs = savedInstanceState.getLongArray(TO_SEND);
+
+            // Add optional authentication params
+            if (savedInstanceState.containsKey(ApplicationConstants.BundleKeys.URL) && savedInstanceState.containsKey(ApplicationConstants.BundleKeys.USERNAME) && savedInstanceState.containsKey(ApplicationConstants.BundleKeys.PASSWORD)) {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java]: Please split into 3 lines to help with readability.

File: collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java [hunk 2]
```diff
@@ -93,14 +96,42 @@ protected void onCreate(Bundle savedInstanceState) {
...
+                username = savedInstanceState.getString(ApplicationConstants.BundleKeys.USERNAME);
+                password = savedInstanceState.getString(ApplicationConstants.BundleKeys.PASSWORD);
+            }
         } else {
             // get instances to upload...
             Intent intent = getIntent();
+            Bundle bundle = intent.getExtras();
             selectedInstanceIDs = intent.getLongArrayExtra(FormEntryActivity.KEY_INSTANCES);
+
+            if (bundle != null && bundle.containsKey(ApplicationConstants.BundleKeys.URL) && bundle.containsKey(ApplicationConstants.BundleKeys.USERNAME) && bundle.containsKey(ApplicationConstants.BundleKeys.PASSWORD)) {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java]: Please split into 3 lines

File: collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java [hunk 3]
```diff
@@ -116,6 +147,11 @@ protected void onCreate(Bundle savedInstanceState) {
             showDialog(PROGRESS_DIALOG);
             instanceServerUploader = new InstanceServerUploader();
 
+            // URL, username & password should always be available in case we are to override the default settings
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java]: I don't understand this comment!

File: collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java [hunk 4]
```diff
@@ -1,5 +1,4 @@
 /*
- * Copyright (C) 2009 University of Washington
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java]: This should stay. You can add an `@author` line to indicate the copyright for your additions. Also, commit history is used to identify ownership.

File: collect_app/src/main/java/org/odk/collect/android/tasks/InstanceServerUploader.java [hunk 5]
```diff
@@ -550,9 +552,9 @@ private boolean processChunk(int low, int high, Outcome outcome, Long... values)
                     String id = c.getString(c.getColumnIndex(InstanceColumns._ID));
                     Uri toUpdate = Uri.withAppendedPath(InstanceColumns.CONTENT_URI, id);
 
-                    // Use the app's configured URL unless the form included a submission URL
+                    // Use the provided submission url OR the submission URL included in the form OR the app's configured URL in that order of priority/precedence
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/tasks/InstanceServerUploader.java]: This would be easier to read as a bulleted list, I think. Otherwise you can just wrap at around 80 chars.

**Review by lognaturel — COMMENTED** [review]
👍 

How does this work with servers that allow anonymous access? Does passing in the URL and empty strings for username and password work?

File: collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java (hunk 4 above)
Comment (ekigamba) [reply collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java]: Sorry, it seems I removed it by mistake. Thanks for that

File: collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java (hunk 1 above)
Comment (ekigamba) [reply collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java]: > How does this work with servers that allow anonymous access? Does passing in the URL and empty strings for username and password work?

Yes, it works if the username & password are empty strings
I tested with https://opendatakit.appspot.com but it seems to accept any username/password. Any other server that you would recommend me to test with?

Comment (lognaturel) [reply collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java]: Awesome. That's the only special case I can think of!

**Comment by lognaturel** [comment]
For QA (CC @kkrawczyk123): the goal here is to make an external app able to trigger sending forms and to optionally provide a destination and destination credentials that override the ones set in Collect.

https://github.com/grzesiek2010/collectTester/pull/22 should let you set a url/username/password in the settings to to try that functionality. It should work with all three set or with just a url. It should override what's in settings and it should override a submission URL set in the form.

Regression risks are around form sending to a server: sending with a URL but no username/password set in settings, sending with credentials, override to submission URL in form, etc.

**Comment by kkrawczyk123** [comment]
- there is no Settings options on Androids 4.1 and 4.2. 

- a password is not hidden on General Settings view. 

- a back arrow from Settings doesn't work on android 7, 6, 5.1 and 4.4  so it works only on 8.1 

I tried different cases:
- input url of aggregate for which credentials are required and didn't input credentials - submission was sent to aggregate from collect's app.

- input url of aggregate for which credentials are not required and didn't input credentials - submission was sent to aggregate from collect's app.

- after input credentials from both cases above submission was sent to a proper aggregate instance. 

- I think that those credentials are stored somehow after removing it as I tried to url with required credentials with username only, url with required credentials with password only and url with required credentials without credentials at all in all cases submission has been sent to the aggregate. 

- then when I tried with aggregate for which credentials are not required submission has been sent to proper aggregate instance on Android 8 and 7, but successfully sent to a proper instance on Android 6.

- when I tried with local instance of aggregate (in cases above I used gae instance) I input only url (credentials are required for that aggregate) I was actually asked for credentials to input when tried to sent it. I input them and sent has been successful. Those credentials were not saved in settings like they are in collect's app in the same situation. 

- I have also encountered a situation when I am asked for credentials, I set them but submissions is sent to demo instance selected as default in collect's app. 

- I input credentials, sent submission, removed credentials, changed to aggregate with required credentials and sent a submission and it was sent to aggregate but it shouldn't.

@lognaturel, I am a bit confused with my results because it looks a bit random for different Androids so it might be not that easy to reproduce by developers, but definitely, I encountered some behaviors which I shouldn't.

@opendatakit-bot unlabel "needs testing"

**Comment by ekigamba** [comment]
@kkrawczyk123 

Are these issues on ODK Collect or on the ODK Collect Tester app?

> there is no Settings options on Androids 4.1 and 4.2.
> a password is not hidden on General Settings view.
> a back arrow from Settings doesn't work on android 7, 6, 5.1 and 4.4 so it works only on 8.1


For instances where the url is blank, the settings on ODK Collect are used. When you enter the url but do not enter the username and password, this is telling ODK Collect that you are an anonymous user.

@kkrawczyk123 Does this explanation clear the any/some of issues indicated above?

**Comment by kkrawczyk123** [comment]
@ekigamba you are right that I should precise more that those issues are connected to Tester app not collect app.

> For instances where the url is blank, the settings on ODK Collect are used. When you enter the url but do not enter the username and password, this is telling ODK Collect that you are an anonymous user.

I can see your point but shouldn't it depend on aggregate settings? I mean I understand that collect consider it as anonymous access but it shouldn't as I don't allow for anonymous access on aggregate and this is what I meant by "aggregate with required credentials". Settings on aggregate are changeable anytime so shouldn't it be recognized somehow?

**Comment by ekigamba** [comment]
@kkrawczyk123 Fixed the issues on both the ODK Collect tester app and ODK Collect. Thanks and get back in case of anything else

**Comment by kkrawczyk123** [comment]
@ekigamba I tested again and still some issues are visible for me:

- I input to collect Tester url of aggregate with required credentials and did not input credentials - only url - submission has been sent to aggregate from collet's app so to demo server (the same result for url+username and url+password). Then I input only url to aggregate which allows for anonymous access - submission has been sent to demo server too. When I input credentials for both, submissions have been sent to a proper aggregate instance. Then I removed credentials for aggregate with anonymous access allowed - submission has been sent successfully. After that when I try to send a submission to aggregate with anonymous access not allowed I saw Server Requires Authentication dialog which I expected in the first try too. So it looks like you have to input some credentials first then the feature actually works as expected. I obtained the same results on Androids 4.1, 4.2, 4.4, 5.1, 6.0, 7.0 and 8.1.  @ekigamba could you please investigate that? 

- I input to Collect's app url of aggregate where anonymous access is not allowed (just url) and left settings on tester app empty. I tried to send a submission and I saw dialog message: 
screenshot_20180822-100541
Shouldn't I see here dialog?: 
screenshot_20180822-100821
which actually would give me the possibility to input username and password and send a submission. It would be good to have error messages unified. @lognaturel, please confirm if you agree with me (or not :) )

- It would be also super user-friendly if credentials input on Server Requires Authentication would be saved on collect'sTester settings like it does in collect's app.

- The new problem occurred on  Androids 4.1, 4.2 and 4.4, the font is white on Form List, Instances List, Submit instance on collectTester app. Take a very careful look on attached screenshot: 
screenshot_2018-08-22-11-52-17

- The password is hidden but it doesn't show the real amount of letters in the password. It's not a high priority issue but it might be a little confusing as it looks like all passwords are consisted of 6 letters.

@opendatakit-bot unlabel "needs testing"

**Comment by ekigamba** [comment]
Thanks @kkrawczyk123 

So I have gone through all the issues and I now completely understand the issues you faced

> I input to collect Tester url of aggregate with required credentials and did not input credentials - only url - submission has been sent to aggregate from collet’s app so to demo server (the same result for url+username and url+password). Then I input only url to aggregate which allows for anonymous access - submission has been sent to demo server too. When I input credentials for both, submissions have been sent to a proper aggregate instance. Then I removed credentials for aggregate with anonymous access allowed - submission has been sent successfully. After that when I try to send a submission to aggregate with anonymous access not allowed I saw Server Requires Authentication dialog which I expected in the first try too. So it looks like you have to input some credentials first then the feature actually works as expected. I obtained the same results on Androids 4.1, 4.2, 4.4, 5.1, 6.0, 7.0 and 8.1. @ekigamba could you please investigate that?

I understand this now. So, for you to specify an anonymous user, right now, you have to specify the username & password as an empty string - This is achieved by just clicking on the username field in settings and leaving it as it is, same with the password field.

> I input to Collect’s app url of aggregate where anonymous access is not allowed (just url) and left settings on tester app empty. I tried to send a submission and I saw dialog message:

This is ok also. So, what this means is that the tester app url is an empty string which is still valid for a testing app and thus is passed together with all the other empty username && password. Empty url is equals to a null hostname. This is behaviour that is recoverable from from the external app

> It would be also super user-friendly if credentials input on Server Requires Authentication would be saved on collect’sTester settings like it does in collect’s app.

This would not be ideal since the user is supposed to be using two different accounts. ODK Collect contains the primary account while other external apps have their own configurations which should not be saved a primary settings in ODK Collect

> The new problem occurred on Androids 4.1, 4.2 and 4.4, the font is white on Form List, Instances List, Submit instance on collectTester app. Take a very careful look on attached screenshot:

This seems to be an issue in previous versions of collectTester. Seems to have escaped everyone :smile:

> The password is hidden but it doesn’t show the real amount of letters in the password. It’s not a high priority issue but it might be a little confusing as it looks like all passwords are consisted of 6 letters.

This follows ODK Collect pattern. The asterisks I have used on collectTester are 2 or 3 less compared to ODK Collects. :sweat_smile:  Thanks you caught it, I wouldn’t have found it as a user issue. Do you really think we should change it?

Most of the issues you faced are in the way I set up the collectTester to work, a little on the ODK Collect implementation.

For ODK Collect implementation side, ODK Collect falls back to the default settings if either the *url, username or password* are null or not provided. For you to completely override the app settings and form settings, you have to provide all three, otherwise they are discarded. I thought this seems like the best way to provide the ability to optionally override current settings

**Comment by lognaturel** [comment]
> you have to specify the username & password as an empty string

This seems really weird to me and it threw me off as well when I was trying it out. It seems natural to want to specify a `URL` extra without credentials. Instead of checking if all 3 are set, could you check to see if there's a `URL`, if so, use that and then in that block also check for the credential extras? I see you said "I thought this seems like the best way to provide the ability to optionally override current settings" but could you explain why?

> I input to Collect's app url of aggregate where anonymous access is not allowed (just url) and left settings on tester app empty. I tried to send a submission and I saw dialog message:

Agreed this is ok.

> It would be also super user-friendly if credentials input on Server Requires Authentication would be saved on collect'sTester settings like it does in collect's app.

Since the app is just for testing, I don't think we really need to think about user friendliness too much. As @ekigamba describes, in a real usage the calling app would maintain credentials.

> The new problem occurred on Androids 4.1, 4.2 and 4.4, the font is white on Form List, Instances List, Submit instance on collectTester app. Take a very careful look on attached screenshot:

Filed at https://github.com/grzesiek2010/collectTester/issues/23

> The password is hidden but it doesn't show the real amount of letters in the password. It's not a high priority issue but it might be a little confusing as it looks like all passwords are consisted of 6 letters.

This is pretty common -- giving away password length is never a great idea.

I was getting confused by what exactly collectTester was doing so I also put together https://github.com/lognaturel/submitInstanceIntent/blob/master/app/src/main/java/com/example/ln/submitinstanceintent/MainActivity.java

File: collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java [hunk 6]
```diff
@@ -93,14 +97,49 @@ protected void onCreate(Bundle savedInstanceState) {
...
+                username = savedInstanceState.getString(ApplicationConstants.BundleKeys.USERNAME);
+                password = savedInstanceState.getString(ApplicationConstants.BundleKeys.PASSWORD);
+            }
         } else {
             // get instances to upload...
             Intent intent = getIntent();
+            Bundle bundle = intent.getExtras();
             selectedInstanceIDs = intent.getLongArrayExtra(FormEntryActivity.KEY_INSTANCES);
+
+            if (bundle != null && bundle.containsKey(ApplicationConstants.BundleKeys.URL)
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java]: When I say "Instead of checking if all 3 are set" in my latest comment, I mean here and when reading from the `savedInstanceState`.

File: collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java [hunk 7]
```diff
@@ -116,6 +155,12 @@ protected void onCreate(Bundle savedInstanceState) {
             showDialog(PROGRESS_DIALOG);
             instanceServerUploader = new InstanceServerUploader();
 
+            // The 3 details - custom URL, username & password passed through the intent should be available if we are to override the default settings
+            // If one of them lacks (in this case it's a null) then we do not override the form or device settings during submission
+            if (url != null && username != null && password != null) {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java]: Oh, and I guess everywhere, really. It looks like this was a very deliberate choice so would be good to understand your reasoning.

**Comment by ekigamba** [comment]
> I was getting confused by what exactly collectTester was doing so I also put together https://github.com/lognaturel/submitInstanceIntent/blob/master/app/src/main/java/com/example/ln/submitinstanceintent/MainActivity.java

Sorry, 🤦‍♂️  I'll try and communicate much better. Hope it helped. If not, don't hesitate

File: collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java (hunk 7 above)
Comment (ekigamba) [reply collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java]: Yea, it was a deliberate choice before I realized that anonymous logins are possible. I also assumed empty strings for the username and password would do ok for anonymous logins as per this 

>  Does passing in the URL and empty strings for username and password work?

 I'll work on changing that 😄

**Comment by lognaturel** [comment]
Nono, you've communicated fine, I was just having trouble figuring out the the intended behavior with extras not being set vs being set to "". Am interested to hear whether you think requiring all is still the way to go or whether it might be possible to make username and password optional.

**Comment by ekigamba** [comment]
It makes complete sense for anonymous logins to either not pass the username & password or use empty strings. In this case, username & password not being passed (this has to happen for both) is okay by me. However, how do we handle only username or password being passed. Do we fallback to the default settings or assume that it's an anonymous login

**Comment by ekigamba** [comment]
@lognaturel I was also thinking of adding a `DELETE_AFTER_SUBMIT` extra which enables the third party app to ensure that instances are deleted after submission. 

The following are some of the other options on how to achieve that:

1. The third party can override the settings by adding the odk.settings file which requires ODK collect main activity to start.
2. Deleting forms from the 3rd party app - Feels hacky and moves some control of form management to the 3rd party app. This means the 3rd party app(s) have to keep track of form instance management changes.

We could assume this to be a default behaviour since the user uses a third party app explicitly and not ODK Collect. The user in this case does not explicitly interact with ODK Collect and having the form instances after submission **might** not be desirable due to filling up storage on the users device.

Basically, the user performing actions on the third party app might not know that this feature is available and might not be in a position to perform the deletions manually.

**Comment by lognaturel** [comment]
> However, how do we handle only username or password being passed. Do we fallback to the default settings or assume that it's an anonymous login

I was thinking that the username and password extras would only be considered if a URL extra exists. That would mean that if only username and password are set, it would be considered an anonymous login. That seems easy enough to document and I don't think it reduces usefulness for clients.

> I was also thinking of adding a DELETE_AFTER_SUBMIT extra which enables the third party app to ensure that instances are deleted after submission.

It makes sense that auto deletion would be desirable when using external apps this way. There's a recently-added feature that lets you configure a form to always auto delete sent instances at the form level. Could that address your need? Sorry, it's not actually documented in XLSForm yet but the description is at https://github.com/opendatakit/docs/issues/837

Adding the intent extra doesn't seem bad, though, it's just annoying to have one more rule to encode since there's already Collect settings which can be overridden by form settings. If the form-level setting doesn't meet your needs, do say so and we can see where to go from there.

**Comment by ekigamba** [comment]
> Adding the intent extra doesn't seem bad, though, it's just annoying to have one more rule to encode since there's already Collect settings which can be overridden by form settings. If the form-level setting doesn't meet your needs, do say so and we can see where to go from there.

Thanks for making it clear that the form `auto_delete` property overrides the device set `auto_delete` setting. It now seems like an easier option. However let me consult about this and get back to you if it meets all our needs.

> I was thinking that the username and password extras would only be considered if a URL extra exists. That would mean that if only username and password are set, it would be considered an anonymous login. That seems easy enough to document and I don't think it reduces usefulness for clients.

We could make the username & password optional so that:

1. URL + Username + Password = Authorisation required login
1. URL = Anonymous logic

How do we handle a URL and either username or password only being provided i.e. URL + username **OR** URL + password. Do we ignore either and assume that its an anonymous login? Do we ignore all extras and default to the app credentials?

**Comment by lognaturel** [comment]
> How do we handle a URL and either username or password only being provided i.e. URL + username OR URL + password. Do we ignore either and assume that its an anonymous login? Do we ignore all extras and default to the app credentials?

Ah, right. I guess not dealing with these cases was a possible reason to make all 3 required. Here's what makes sense to me in terrible pseudocode:
```
if (URL) {
  if (username && password) {
     send to URL with credentials
  }
  send to URL anonymously
}

send using Collect settings
```

The docs would say something like: "Three optional extras are supported: URL, username and password. If none are specified, the credentials in Collect's settings are used. If only the URL is specified, the submission goes to that URL anonymously. To send a submission using credentials, URL, username and password must all be set. If credentials are provided without a URL, they are ignored. If a URL is provided but one of username or password is missing, the submission goes to that URL anonymously."

I don't know, does that seem reasonable?

**Comment by ekigamba** [comment]
Yea, it does. At least that's clear

**Comment by ekigamba** [comment]
@lognaturel Made the changes and tested using the same collectTester version. Get back in case of anything

**Comment by lognaturel** [comment]
The credentials part looks fine to me. @mmarciniak90 would be great to get a quick check with the intents tester app again and a quick regression check.

My hope was that you wouldn't need to add the instance deletion extra and could use the form-level configuration. Why is that one not sufficient?

@yanokwa or @grzesiek2010, while the QA pass is going on, it would be good to get an extra sanity check here, especially on adding the delete after send extra and the names for those new extras since those can't really be changed once accepted. This currently makes it so InstanceUploaderActivity accepts extras with names `instances`, `URL`, `USERNAME`, `PASSWORD` and possibly `DELETE_INSTANCE_AFTER_SUBMISSION`. Note the inconsistent case. 

Different activities already have different conventions so I don't know if it matters here. I do think CONSTANT_CASE is more common than camelCase for extra names so the options are matching `instances` and going with less standard casing or using CONSTANT_CASE and being inconsistent with `instances`. I suppose that if the inconsistency really bothers us we could add a follow-up PR to introduce `INSTANCES` as an alias.

**Comment by ekigamba** [comment]
@lognaturel I talked with the team and it seems that the form submissions will end up taking too much storage space after being submitted with no reason to retain them. The form authors might also not be too tech savy and often forget to add the `auto_delete` for every form.

**Comment by mmarciniak90** [comment]
Hi @ekigamba!
I started testing your changes.

I noticed unexpected behavior
* The internal app allows sending form which was not marked as finalized, which is not allowed on ODK Collect.

My testing is still in progress but I wanted to let you know about this as soon as possible.

**Comment by ekigamba** [comment]
Thanks @mmarciniak90 , I'll fix that

**Comment by ekigamba** [comment]
@mmarciniak90 @lognaturel I fixed ODK Collect tester to only show finalized instances. Do you think I should make a check to make sure that the instanceIds provided are actually finalized on ODK Collect InstanceUploaderActivity for cases where the activity has been started by an external app?

**Comment by lognaturel** [comment]
>  The form authors might also not be too tech savy and often forget to add the auto_delete for every form.

Ok, adding the extra seems ok to me but would like @yanokwa or @grzesiek2010 to take a look and offer an opinion as well.

Good catch on the unfinalized forms, @mmarciniak90. I do think it would be good to have a check now that arbitrary form IDs can be passed in because unfinalized forms can violate constraints, etc. That said, I know this is dragging on so I'd be ok with it being filed as a separate issue. Most external apps will likely want to filter what they allow the user to initiate a send for anyway.

**Comment by yanokwa** [comment]
I think we should use upper case for extras as that's common practice. We can add an upper-case alias for `instances`  in a follow-on PR and eventually deprecate support for the lower case.

I initially didn't love `DELETE_INSTANCE_AFTER_SUBMISSION` because I prefer using consistent naming from the UI down to the code. But @lognaturel explained that it's specific to a single submission and not really auto delete so I'm fine with it.

**Comment by mmarciniak90** [comment]
@ekigamba @lognaturel 
I noticed one case but I'm not sure that it needs to be fixed.
1. The Aggregate server is set on Collect
2. The user uses Tester but does not add URL -> submissions are sent to Aggregate set in Collect
3. The user adds URL in Tester app
4. The user removes URL in Tester app
5. The user is no longer able to send a submission to Aggregate which is set in Collect. `Host name may not be null` error is visible.

I expected that submission will be sent to Aggregate which is set in Collect. 

The second thing, that I'm not sure about is the information dialog.
screenshot_2018-09-13-13-26-56
Is it expected to always see this Aggregate URL: https://opendatakit.appspot.com?

**Comment by ekigamba** [comment]
@mmarciniak90 Thats a case on the Tester that would need changing. The behaviour is ok, since when you removed the url you basically left it as an empty string. Empty urls have no hosts thus the reason it gave you that error. If Collect fell back to the default settings, that would mean submissions would go to the wrong server instead of actually failing so that the 3rd party/external app can fix the issue on their end

**Comment by mmarciniak90** [comment]
@ekigamba So maybe it is a good idea that it never let to send a submission if the URL field is empty in the Tester.

**Comment by ekigamba** [comment]
Probably let tester pass any incorrect configurations so that we can tell if ODK Collect can handle them correctly OR as we expect it to. 🤷‍♂️ 

What do you think?

**Comment by lognaturel** [comment]
> The second thing, that I'm not sure about is the information dialog.

The intent is for it to show an example. A fixed URL is fine. 

I agree that the case with a null URL is as expected and that no changes are needed. As @ekigamba says, it’s best for the tester not to do too much sanitizing or it could hide things Collect can’t handle. 

Good things to check, @mmarciniak90!

**Comment by mmarciniak90** [comment]
@ekigamba, @lognaturel thanks for responses! Your explanations are clear to me. 
So, it can be to merge. 

Verified on Android 5.1, 6.0, 7.0, 8.1
It was problematic to verify on Android 4.4, 4.2, 4.1 because of https://github.com/grzesiek2010/collectTester/issues/23 but it looks good too

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

**Review by lognaturel — APPROVED** [review]
